storage/src/vespa/storage/distributor/distributor.cpp

```diff
@@ -321,6 +330,43 @@ bool should_be_handled_by_top_level_bucket_db_updater(const api::StorageMessage&
     }
 }
 
+document::BucketId
+get_bucket_id_for_striping(const api::StorageMessage& msg, const DistributorNodeContext& node_ctx)
+{
+    if (!msg.getBucketId().isSet()) {
+        // Calculate a bucket id (dependent on the message type) to dispatch the message to the correct distributor stripe.
+        switch (msg.getType().getId()) {
+            case api::MessageType::PUT_ID:
+            case api::MessageType::UPDATE_ID:
+            case api::MessageType::REMOVE_ID:
+                return node_ctx.bucket_id_factory().getBucketId(dynamic_cast<const api::TestAndSetCommand&>(msg).getDocumentId());
+            case api::MessageType::REQUESTBUCKETINFO_REPLY_ID:
+            {
+                const auto& reply = dynamic_cast<const api::RequestBucketInfoReply&>(msg);
+                if (!reply.getBucketInfo().empty()) {
+                    // Note: All bucket ids in this reply belong to the same distributor stripe, so we just use the first entry.
+                    return reply.getBucketInfo()[0]._bucketId;
+                } else {
+                    return reply.getBucketId();
+                }
+            }
+            default:
+                return msg.getBucketId();
+        }
```

Comment:
Handling of visitor-related requests will probably have to be added here, since e.g. `CreateVisitorCommand` does not have an explicit single bucket ID, and they must be routed to the stripe that owns the superbucket.